spec/formatting.md

```diff
@@ -18,3 +18,71 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided to the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is invalid.
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable** errors occur when a variable reference cannot be resolved.
+
+- **Selection errors** cover failures encountered during selection.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
```

Comment:
What are the cases where this can occur? I would assume that a given key either matches or does not match a given selector value. The only selection error that can occur is if there is a missing fallback... which is already covered? What am I forgetting?

I think it more likely that selector errors would be things like type mismatches (trying to run a plural selector on a string), in which case the description would be more like:

> **Selector errors** are selector internal errors, such as type mismatches or implementation defined internal errors.

Reply:
Something like `{(horse) :plural}` could fail as a selector, the same way that e.g. `{(horse) :number}` could fail when formatting. There's also potential for e.g. internally inconsistent options bags, or any other sort of weirdness. But sure, type mismatches are among the likeliest. Do you think that ought to be specifically highlighted?

Reply:
This is meant to cover errors thrown in the custom code that's needed to implement a selector like `:plural`. For example, in JS this means that somewhere we'll need to have a call like:
```js
new Intl.PluralRules(lc, opt).select(num) // returns 'zero', 'one', ..., or 'other'
```
So while it's unlikely that this user-controlled code will throw an error, that's still possible. If it happens, we should report it as a Selector Error.

Reply:
Plural is a good test case because the features of a plural *formatter* go beyond what `Intl.PluralRules` provides. For example, a plural can have `=n` cases (such as `=0` or `=1`) which are not handled by plural rules in JS. The plural rules `select()` is only a part of how the selector itself operates (another case would be if the locale returned e.g. the `many` keyword but there is no `many` selector: this is *not* an error as the `*` case should match it--this happens all the time when the message has not yet been translated!!)

Reply:
How do you feel about a "internal function error" that would cover all errors in function implementations, formatter or selector?

Reply:
I would prefer keeping resolution, selection and formatting errors as distinguishable from each other. Where available, I could see them all inheriting from some common "internal function error" type, though.

Reply:
Marking as resolved; the PR now includes illustrative examples and questions raised in this thread have been answered.